docs/design/sql/08-expressions.md

```diff
@@ -0,0 +1,192 @@
+# SQL Expressions
+
+## Overview
+
+Expressions is one of the core elements of every SQL implementation.
+Expressions can participate in projections (`SELECT a + b FROM t`) and
+filters (`SELECT * FROM t WHERE a + b < 10`). Expressions may be roughly
+subdivided into 4 general categories:
+
+- Terminal expressions not referring to other expressions: column,
+  literal, parameter and type expressions.
```

Comment:
What are type expressions?

Reply:
`CAST(1 AS INT)`, `INT` is referring to a type, clarified in d1d2929413835f7a462bc88686630dd57203b27c

Reply:
But these can refer to other expressions, cant' they? You can do `CAST(1+column1+UPPER(col2)) as INT`. So they don't belong to this category.

Reply:
I meant the `INT` part itself, it provides a reference to a type in expressions, in Calcite it's represented as `SqlIdentifier` which is a subclass of `SqlNode`, so it's a regular AST node.

Reply:
Makes sense this way, I never thought of the `INT` as of a _type expression_. I'll suggest an alternative text to current version though.